include/sys/dmu_objset.h

```diff
@@ -125,6 +126,12 @@ struct objset {
 	kmutex_t os_user_ptr_lock;
 	void *os_user_ptr;
 	sa_os_t *os_sa;
+
+	/* kernel thread to upgrade this dataset */
+	kthread_t *os_upgrade_thread;
+	kmutex_t os_upgrade_lock;
+	kcondvar_t os_upgrade_cv;
+	boolean_t os_upgrade_exit;
```

Comment:
What do you think about placing this generic upgrade code in the `spa_t` instead.  I'm concerned about spawning a huge number of threads for existing users who commonly mount 1000's of datasets and are upgrading.  By placing the thread in the `spa_t` this could be serialized.  Or even better a taskq with say 1 worked thread per-cpu could be used to get a reasonable amount of parallelism.